### PR TITLE
[#358] pool: Start even if not all node healthy

### LINKED ISSUE #358: Pool requires all endpoints to be up
I tried to start S3 gateway with pool of 2 nodes, where one node is down.
```yaml
peers:                                                                                                                                                                                              
  0:                                                                                                                                                                                                
    address: s01.neofs.devenv:8080 # <- This node is down                                                                                                            
  1:      
    address: s02.neofs.devenv:8080
```

S3 gateway didn't start. Don't think it is expected behavior for the pool during initialization.

```
fatal   s3-gw/app.go:271        failed to dial connection pool  {"error": "client dial: open gRPC connection: gRPC dial: context deadline exceeded"}
main.getPool
        github.com/nspcc-dev/neofs-s3-gw/cmd/s3-gw/app.go:271
main.newApp
        github.com/nspcc-dev/neofs-s3-gw/cmd/s3-gw/app.go:90
main.main
        github.com/nspcc-dev/neofs-s3-gw/cmd/s3-gw/main.go:15
runtime.main
        runtime/proc.go:250
```

### PR DESCRIPTION
**Pull request by KirillovDenis**
close #358
Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: pool/pool.go [hunk 1]
```diff
@@ -1404,22 +1543,26 @@ func (p *Pool) Dial(ctx context.Context) error {
 	for i, params := range p.rebalanceParams.nodesParams {
 		clients := make([]client, len(params.weights))
 		for j, addr := range params.addresses {
-			c, err := p.clientBuilder(ctx, addr)
-			if err != nil {
-				return err
+			c := p.clientBuilder(addr)
+			if err := c.dial(ctx); err != nil {
```
Comment (cthulhu-rider) [inline pool/pool.go:1547]: Shouldn't we `continue`?

File: pool/pool.go [hunk 2]
```diff
@@ -91,6 +95,9 @@ type clientStatus interface {
 	methodsStatus() []statusSnapshot
 }
 
+// ErrPoolClientUnhealthy is an error to indicate that client in pool is unhealthy.
+var ErrPoolClientUnhealthy = errors.New("pool client unhealthy")
```
Comment (cthulhu-rider) [inline pool/pool.go:99]: Do we need to handle this error outside the package? If not, do not export it.

Comment (KirillovDenis) [reply pool/pool.go:99]: In theory we can do retry in app code if  such error happened

File: pool/pool.go [hunk 3]
```diff
@@ -248,44 +256,107 @@ func (x *wrapperPrm) setResponseInfoCallback(f func(sdkClient.ResponseMetaInfo)
...
+	if _, err := cl.EndpointInfo(ctx, sdkClient.PrmEndpointInfo{}); err != nil {
+		c.setUnhealthy()
+		return false, wasHealthy
+	}
+
+	c.setHealthy()
+	return true, !wasHealthy
+}
+
+func (c *clientWrapper) getClient() (*sdkClient.Client, error) {
```
Comment (alexvanin) [inline pool/pool.go:339]: `getClient` is introduced to solve potential races with `c.client` access?

File: pool/pool.go [hunk 4]
```diff
@@ -248,44 +256,107 @@ func (x *wrapperPrm) setResponseInfoCallback(f func(sdkClient.ResponseMetaInfo)
...
-	return res, nil
+	return nil
+}
+
+// restartIfUnhealthy checks healthy status of client and recreate it if status is unhealthy.
+// Return current healthy status and indicating if status was changed by this function call.
+func (c *clientWrapper) restartIfUnhealthy(ctx context.Context) (healthy, changed bool) {
+	var wasHealthy bool
+	if _, err := c.endpointInfo(ctx, prmEndpointInfo{}); err == nil {
+		return true, false
```
Comment (alexvanin) [inline pool/pool.go:304]: This return assumes that `wasHealthy` was true (changed == false). Is it expected?

File: pool/pool.go (hunk 2 above)
Comment (cthulhu-rider) [reply pool/pool.go:99]: We should either:
* wait when theory becomes reality, or
* mention this error in docs of all methods which can return it

File: pool/pool.go (hunk 3 above)
Comment (KirillovDenis) [reply pool/pool.go:339]: Yes. We recreate client sometimes

File: pool/pool.go (hunk 4 above)
Comment (KirillovDenis) [reply pool/pool.go:304]: Yes. If client wasn't healthy the `c.endpointInfo` returns error

**Comment by alexvanin** [comment]
@KirillovDenis can we run some http or s3 tests with new pool before merge? Some experimental branch which then will be merged into `support/v0.25`

**Review by cthulhu-rider — APPROVED** [review]

**Comment by KirillovDenis** [comment]
> Some tests will be appreciated before merge.

Tested on http-gw and s3-gw with @dansingjulia

**Pull request merged**